README.md
# create-sky

### Mysql 数据结构输出模板

> 问答如下

- ? 请选择工程模板 > Choice your Template: DB-report > Mysql 库表结构生成
- ? .项目路径,相对当前 [.]: .
- ? Mysql host [127.0.0.1]: 127.0.0.1
- ? Mysql port [3306]: 3306
- ? Mysql user [root]: root
- ? Mysql password [123456]: [hidden]
- ? Mysql database [test]: test


CHANGELOG.md
## [1.0.3](https://github.com/kongnet/create-sky/compare/v1.0.2...v1.0.3) (2020-08-11)




## [1.0.2](https://github.com/kongnet/create-sky/compare/v1.0.1...v1.0.2) (2020-08-11)




## 1.0.1 (2020-08-11)






template/db-report/db-report.js
const SkyDB = require('j2sql2')
const fs = require('fs')
const path = require('path')
const $ = require('meeko')
const dbReport = JSON.parse(
  fs
    .readFileSync('db-report.json')
    .toString()
    .replaceAll('\r\n', '')
    .replaceAll(' ', '')
)
//console.log(dbReport)
async function initDB () {
  const skyDB = new SkyDB({
    mysql: dbReport
  })
  const db = await skyDB.mysql // 创建mysql实例
  global.db = db
}
let tableSql = `SELECT
  table_schema as db,
  table_name as table_name,
  table_comment as table_comment
  FROM
  information_schema.\`TABLES\`
  WHERE
  table_schema NOT IN ( 'information_schema', 'performance_schema', 'mysql', 'sys' );`
let columnSql = `SELECT
  table_schema AS db_name,
  table_name AS table_name,
  column_name AS column_name,
  column_type AS column_type,
  column_comment AS column_comment
  FROM
  information_schema.\`COLUMNS\`
  WHERE
  table_schema NOT IN ( 'information_schema', 'performance_schema', 'mysql', 'sys' )
  ORDER BY table_schema;`

async function allColumnList () {
  await initDB()
  let n = 1
  let r = await db.run(columnSql)
  let tableArr = r
    //.filter(item => item['db_name'].indexOf('ft') === 0) 可过滤
    .map(x => {
      return [
        n++,
        x.db_name,
        x.table_name,
        x.column_name,
        x.column_type,
        x.column_comment
      ]
    })

  let htmlStr = $.tools.genTemp.genHtml(
    '',
    $.tools.genTemp.gridTable(
      [
        {
          dataTitleArr: ['序号', '数据库名', '表名', '列名', '列类型', '注释'],
          dataArr: tableArr,
          dataTitle: '数据库表结构'
        }
      ],
      'open'
    )
  )
  fs.writeFileSync('mysql-output.html', htmlStr)
}

async function tableColumnList () {
  await initDB()
  let tableResult = await db.run(tableSql)
  let tableComment = []
  tableResult.forEach(x => {
    tableComment[x.db + '=>' + x.table_name] = {
      db: x.db,
      comment: x.table_comment
    }
  })

  let n = -1
  let r = await db.run(columnSql)
  let tableDiffName = ''

  let htmlStr = ''
  let columnArr = []
  let tableArr = r
    //.filter(item => item['db_name'].indexOf('ft') === 0) //可过滤
    .map(x => {
      if (tableDiffName !== x.table_name) {
        n++
        columnArr.push([])
        if (n > 0) {
          let dbTableName = columnArr[n - 1][0][0] + '=>' + tableDiffName
          htmlStr += $.tools.genTemp.gridTable(
            [
              {
                dataTitleArr: ['数据库名', '表名', '列名', '列类型', '注释'],
                dataArr: columnArr[n - 1],
                dataTitle:
                  dbTableName + '（' + tableComment[dbTableName].comment + '）'
              }
            ],
            'open'
          )
        }
        tableDiffName = x.table_name
      } else {
        //do nothing
      }
      columnArr[n].push([
        x.db_name,
        x.table_name,
        x.column_name,
        x.column_type,
        x.column_comment
      ])
    })

  htmlStr = $.tools.genTemp.genHtml('', htmlStr)
  fs.writeFileSync('mysql-output.html', htmlStr)

  /* 可以打包成一个exe程序！
  // pkg打包无法打入puppeteer
  const browser = await puppeteer.launch()
  const page = await browser.newPage()
  await page.setContent(htmlStr, {
    waitUntil: 'domcontentloaded'
  })
  await page.pdf({ path: './mysql-output.pdf', format: 'A4' })
  await browser.close()
  */
  process.exit(0)
}
async function mainTest () {
  //测试函数
  await allColumnList()
  //await tableColumnList()
}
//mainTest()
module.exports = {
  allColumnList,
  tableColumnList,
  dbConfig: dbReport
}


template/db-report/task.js
const $ = require('meeko')
const fs = require('fs')
const path = require('path')
const inquirer = require('inquirer')
const gen = require('j2dir')
const util = require('util')
const outObj = require('./output-file.js')
const cp = require('child_process')
const mysqlDefaultOption = {
  host: '127.0.0.1',
  port: '3306',
  user: 'root',
  password: '123456',
  database: 'test',
  pool: 10,
  timeout: 60000,
  charset: 'utf8mb4',
  supportBigNumbers: true,
  multipleStatements: true,
  connectionLimit: 100,
  showSql: true
}
async function main () {
  let r = await inquirer.prompt([
    {
      name: 'project-path',
      message: $.c.c('.项目路径,相对当前 [.]: '),
      type: 'input',
      default: '.'
    },
    {
      name: 'host',
      message: $.c.c('Mysql host [127.0.0.1]: '),
      type: 'input',
      default: '127.0.0.1'
    },
    {
      name: 'port',
      message: $.c.c('Mysql port [3306]: '),
      type: 'input',
      default: '3306'
    },
    {
      name: 'user',
      message: $.c.c('Mysql user [root]: '),
      type: 'input',
      default: 'root'
    },
    {
      name: 'password',
      message: $.c.c('Mysql password [123456]: '),
      type: 'password',
      default: '123456'
    },
    {
      name: 'database',
      message: $.c.c('Mysql database [test]: '),
      type: 'input',
      default: 'test'
    }
  ])
  let fileObj = outObj()
  if (r['project-path'] !== '.') {
    try {
      fs.mkdirSync(path.join(process.cwd(), r['project-path']))
    } catch (e) {
      //do nothing
    }
  }
  for (i in fileObj) {
    fs.copyFileSync(
      fileObj[i][0],
      path.join(process.cwd(), r['project-path'], i)
    )
  }

  let optObj = Object.assign(mysqlDefaultOption, r)
  fs.writeFileSync(
    path.join(process.cwd(), r['project-path'], 'db-report.json'),
    JSON.stringify(optObj)
  )
  console.log('Mysql 库表结构模板生成完成！')

  cp.execSync('npm i --registry=https://registry.npm.taobao.org', {
    cwd: r['project-path']
  })
  cp.execSync('node index', { cwd: r['project-path'] })
  console.log($.c.m('mysql-output.html'), '已经生成')
}
main()
